### PR TITLE
Node upgrades docs

### PR DESCRIPTION
**Pull request by bogdan-rosianu**
#### Description of the pull request (what is new / what has changed)
documented the activation epoch mechanism and how nodes upgrades work

#### Did you test the changes locally ?
- [x] yes
- [ ] no

#### Which category (categories) does this pull request belong to?
- [x] document new feature
- [x] update documentation that is not relevant anymore
- [ ] add examples or more information about a component
- [ ] fix grammar issues
- [ ] other

### CODE REVIEW

File: docs/validators/node-upgrades.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+As compared to other protocols that perform upgrades that start at a specific block height, releases for Elrond nodes
+don't have a specific block height where the new updates become effective, but rather the first block in the 
+activation epoch will make the nodes proceed with the updated versions of the components.
+
+Since the height of the first block in an epoch isn't deterministic (due to possible roll-backs), the Network Height
+where a feature becomes effective cannot be calculated.
+
+However, the time when a new feature of a bugfix becomes effective is deterministic, as epochs have fixed lengths in rounds.
+Currently, Elrond Mainnet has epochs of `14,400` rounds and a round is `6 sec`. This results in a `24h` epoch.
```
Comment (iulianpascalau) [inline docs/validators/node-upgrades.md:43]: actually is 14401 :D

File: docs/validators/node-upgrades.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+### *Activation epoch example*
+
+For example, let's say that we want to introduce a feature so that smart contracts can receive a `PayableBySC` metadata that
+will allow them to receive EGLD or other tokens from other smart contracts. 
+
+*Timeline example*
+- the Elrond Mainnet is at epoch `590`.
+- currently, the node binary doesn't know about the `PayableBySC` metadata so if one wants to try it, an error like `invalid metadata` 
+will be returned.
+- at epoch `600`, we release a new node binary that contains the `PayableBySC` metadata that will become activate starting with epoch `613`.
```
Comment (iulianpascalau) [inline docs/validators/node-upgrades.md:54]: that will become active *

File: docs/validators/node-upgrades.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+- all nodes operators perform the upgrade.
+- when epoch `613` begins, that metadata will be activated and smart contracts can now have that flag.
+- if one wants to issue a smart contract that is `PayableBySC`, it will work.
+
+- nodes that didn't perform the upgrade will produce a different output of the transaction (as compared to the majority)
```
Comment (iulianpascalau) [inline docs/validators/node-upgrades.md:59]: and won't be able to keep up with the rest of the chain * 
this sounds better as a node might be an observer and not participate in the consensus

File: docs/validators/node-upgrades.md (hunk 1 above)
Comment (bogdan-rosianu) [reply docs/validators/node-upgrades.md:43]: rephrased

File: docs/validators/node-upgrades.md (hunk 2 above)
Comment (bogdan-rosianu) [reply docs/validators/node-upgrades.md:54]: done

File: docs/validators/node-upgrades.md (hunk 3 above)
Comment (bogdan-rosianu) [reply docs/validators/node-upgrades.md:59]: done

File: docs/validators/node-upgrades.md [hunk 4]
```diff
@@ -39,8 +39,9 @@ activation epoch will make the nodes proceed with the updated versions of the co
 Since the height of the first block in an epoch isn't deterministic (due to possible roll-backs), the Network Height
 where a feature becomes effective cannot be calculated.
 
-However, the time when a new feature of a bugfix becomes effective is deterministic, as epochs have fixed lengths in rounds.
-Currently, Elrond Mainnet has epochs of `14,400` rounds and a round is `6 sec`. This results in a `24h` epoch.
+However, the time when a new feature of a bugfix becomes effective can be calculated, as epochs have fixed lengths in rounds.
+Currently, Elrond Mainnet has epochs of `14,400` rounds and a round is `6 sec`. This results in a `24h` epoch. However,
+there can delays of a few rounds, due to roll-backs.
```
Comment (iulianpascalau) [inline docs/validators/node-upgrades.md:44]: there can be delays *
and due to rollbacks of the start of epoch metablocks

Comment (bogdan-rosianu) [reply docs/validators/node-upgrades.md]: done

**Review by iulianpascalau — APPROVED** [review]

File: docs/validators/node-upgrades.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
+---
+id: node-upgrades
+title: Elrond Node upgrades
+---
+
+## **Introduction**
+
+Once a new node's binary is ready to be deployed on one of a network (mainnet, testnet or devnet), nodes operators must
```
Comment (andreibancioiu) [inline docs/validators/node-upgrades.md:8]: `one of the networks`?

File: docs/validators/node-upgrades.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+plus via other communication channels, depending on the case. 
+
+## **Types of upgrades**
+
+Currently, we have the following types of upgrades:
+
+A. - **all nodes need to upgrade**:  upgrades that involve processing changes with an activation epoch (as explained below) 
+and have to be performed by all nodes operators in order to keep the same view over the network and not cause service disruptions.
+
+B. - **optional upgrades**: upgrades that for example simply add a new Rest API endpoint or improve the trie syncing timing
```
Comment (andreibancioiu) [inline docs/validators/node-upgrades.md:19]: `that, for example,`

File: docs/validators/node-upgrades.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+## **Types of upgrades**
+
+Currently, we have the following types of upgrades:
+
+A. - **all nodes need to upgrade**:  upgrades that involve processing changes with an activation epoch (as explained below) 
+and have to be performed by all nodes operators in order to keep the same view over the network and not cause service disruptions.
+
+B. - **optional upgrades**: upgrades that for example simply add a new Rest API endpoint or improve the trie syncing timing 
+are not critical from a processing point of view, and they are optional. If the nodes operators think the new feature will help them,
```
Comment (andreibancioiu) [inline docs/validators/node-upgrades.md:20]: This seems to be a new paragraph?

File: docs/validators/node-upgrades.md [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+they can proceed with the upgrade without losing the compatibility with the network.
```
Comment (andreibancioiu) [inline docs/validators/node-upgrades.md:21]: Some newlines are slightly unexpected (in the markdown source, not necessarily in the rendered output).

File: docs/validators/node-upgrades.md [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+C. - **only validators need to upgrade**: upgrades that, for example, include new features that only trigger validators (ratings changes, 
+transactions selection improvements and so on). Observers (nodes that don't have a stake attached) don't need to perform the upgrade.
```
Comment (andreibancioiu) [inline docs/validators/node-upgrades.md:24]: Perhaps also add `(but can upgrade nonetheless if desired)`.

File: docs/validators/node-upgrades.md [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+## **Activation epochs**
+
+In order to make the upgrades as smooth as possible and to ensure that each node has the same view over the network,
```
Comment (andreibancioiu) [inline docs/validators/node-upgrades.md:28]: Perhaps also add `..., at a given moment`?

File: docs/validators/node-upgrades.md [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+are not critical from a processing point of view, and they are optional. If the nodes operators think the new feature will help them,
+they can proceed with the upgrade without losing the compatibility with the network.
+
+C. - **only validators need to upgrade**: upgrades that, for example, include new features that only trigger validators (ratings changes, 
+transactions selection improvements and so on). Observers (nodes that don't have a stake attached) don't need to perform the upgrade.
+
+## **Activation epochs**
+
+In order to make the upgrades as smooth as possible and to ensure that each node has the same view over the network, 
+Elrond has a so called *activation epoch* mechanism that allows the node to have both versions of the code (before and
```
Comment (andreibancioiu) [inline docs/validators/node-upgrades.md:29]: Suggestion: `to implement both behaviors of the protocol  - the old (current) one, and the new one, planned for activation at a specific epoch`. 

Suggestion: `This mechanisms ensures that, **until the protocol change becomes active**, nodes with an upgraded codebase / binary remain compatible with nodes that did not perform the upgrade, and consensus is held.`

File: docs/validators/node-upgrades.md [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+after the activation epoch) so that we can maintain the backwards compatibility. The system will behave accordingly to 
+the epoch.
+
+### **Deterministic time / height for upgrades**
+
+As compared to other protocols that perform upgrades that start at a specific block height, releases for Elrond nodes
+don't have a specific block height where the new updates become effective, but rather the first block in the 
+activation epoch will make the nodes proceed with the updated versions of the components.
+
+Since the height of the first block in an epoch isn't deterministic (due to possible roll-backs), the Network Height
```
Comment (andreibancioiu) [inline docs/validators/node-upgrades.md:39]: Although the phrasing `isn't deterministic` is correct, perhaps we can rewrite to `cannot be known in advance`?

Comment (andreibancioiu) [reply docs/validators/node-upgrades.md:39]: Also, above we use lower case, but here title case (`Network Height`).

File: docs/validators/node-upgrades.md [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+where a feature becomes effective cannot be calculated.
+
+However, the time when a new feature of a bugfix becomes effective can be calculated, as epochs have fixed lengths in rounds.
+Currently, Elrond Mainnet has epochs of `14,400` rounds and a round is `6 sec`. This results in a `24h` epoch. However,
+there can be delays of a few rounds, due to rollbacks of the start of epoch metablocks.
```
Comment (andreibancioiu) [inline docs/validators/node-upgrades.md:44]: Do we still need this paragraph? (is it related to the rest of the content or might confuse the reader?)

File: docs/validators/node-upgrades.md [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+For example, let's say that we want to introduce a feature so that smart contracts can receive a `PayableBySC` metadata that
+will allow them to receive EGLD or other tokens from other smart contracts. 
+
+*Timeline example*
+- the Elrond Mainnet is at epoch `590`.
+- currently, the node binary doesn't know about the `PayableBySC` metadata so if one wants to try it, an error like `invalid metadata` 
+will be returned.
+- at epoch `600`, we release a new node binary that contains the `PayableBySC` metadata that will become active starting with epoch `613`.
+- all nodes operators perform the upgrade.
+- when epoch `613` begins, that metadata will be activated and smart contracts can now have that flag.
```
Comment (andreibancioiu) [inline docs/validators/node-upgrades.md:57]: `the new feature activates and the new metadata is recognized and accepted etc.`

File: docs/validators/node-upgrades.md [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+- if one wants to issue a smart contract that is `PayableBySC`, it will work.
+
+- nodes that didn't perform the upgrade will produce a different output of the transaction (as compared to the majority)
+and won't be able to keep up with the rest of the chain.
+
+*Backwards compatibility explained*:  
+If one wants to process all the blocks since genesis (via `full archive` or via `import-db`) with the released binary
+it will behave this way:
```
Comment (andreibancioiu) [inline docs/validators/node-upgrades.md:65]: Unexpected newline (in the source code, not in the rendered output), as well.

File: docs/validators/node-upgrades.md [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+and won't be able to keep up with the rest of the chain.
+
+*Backwards compatibility explained*:  
+If one wants to process all the blocks since genesis (via `full archive` or via `import-db`) with the released binary
+it will behave this way: 
+- if for example, in epoch `455` there was a transaction that tried to set the `PayableBySC` metadata, it will process it 
+as `invalid metadata`
+- for transactions in epochs newer than `613` it will process the new metadata.
+
+|                 | Before activation epoch | After activation epoch |
```
Comment (andreibancioiu) [inline docs/validators/node-upgrades.md:70]: Since this is an example, you can write the epoch numbers in parenthesis, in the table header.

**Review by andreibancioiu — APPROVED** [review]

File: docs/validators/node-upgrades.md (hunk 5 above)
Comment (bogdan-rosianu) [reply docs/validators/node-upgrades.md:8]: done

File: docs/validators/node-upgrades.md (hunk 7 above)
Comment (bogdan-rosianu) [reply docs/validators/node-upgrades.md:20]: They will be displayed on the same line on the docs website, but I like to split into multiple lines so I don't have too long lines

File: docs/validators/node-upgrades.md (hunk 9 above)
Comment (bogdan-rosianu) [reply docs/validators/node-upgrades.md:24]: done

File: docs/validators/node-upgrades.md (hunk 11 above)
Comment (bogdan-rosianu) [reply docs/validators/node-upgrades.md:29]: done

File: docs/validators/node-upgrades.md (hunk 13 above)
Comment (bogdan-rosianu) [reply docs/validators/node-upgrades.md:44]: Yes, it's ok

**Review by andreibancioiu — DISMISSED** [review]

File: docs/validators/node-upgrades.md [hunk 17]
```diff
@@ -16,27 +16,29 @@ Currently, we have the following types of upgrades:
...
 ## **Activation epochs**
 
-In order to make the upgrades as smooth as possible and to ensure that each node has the same view over the network, 
-Elrond has a so called *activation epoch* mechanism that allows the node to have both versions of the code (before and 
-after the activation epoch) so that we can maintain the backwards compatibility. The system will behave accordingly to 
-the epoch.
+In order to make the upgrades as smooth as possible and to ensure that each node has the same view over the network at a given moment, 
+Elrond has a so called *activation epoch* mechanism that allows the node to implement both behaviors of the protocol - 
+the old (current) one, and the new one, planned for activation at a specific epoch. This mechanisms ensures that, 
+**until the protocol change becomes active**, nodes with an upgraded codebase / binary remain compatible with nodes that
```
Comment (iulianpascalau) [inline docs/validators/node-upgrades.md:32]: this is desirable but not always attainable
If we want to be politically correct we should state the following:
`This mechanisms ensures that, **until the protocol change becomes active**, nodes with an upgraded codebase / binary remain compatible with nodes that did not perform the upgrade, and consensus is held.
Although this happens in 99.9% cases, this is not 100% guaranteed due to the unforeseen consequences when upgrading the code base in parallel with executing transactions generated by 3rd parties (Elrond blockchain users)`

**Review by iulianpascalau — CHANGES_REQUESTED** [review]

File: docs/validators/node-upgrades.md (hunk 17 above)
Comment (bogdan-rosianu) [reply docs/validators/node-upgrades.md]: added

**Review by andreibancioiu — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**